### PR TITLE
Keyed SipHash

### LINKED ISSUE #22520: Function sipHash{64,128} not usable as MAC (not seedable)
**Describe the unexpected behaviour**
SipHash is designed to be used as MAC (message authentication code) crypto primitive. To be able to do so it takes two arguments: `calcSipHash(msg, key)`. The ClickHouse™ functions `sipHash64` and `sipHash128` function do only expose the first argument `msg` (see [sipHash64 Doc](https://clickhouse.tech/docs/en/sql-reference/functions/hash-functions/#hash_functions-siphash64)).

**How to reproduce**
* See documentation of function sipHash64 and compare with [Wikipedia - SipHash](https://en.wikipedia.org/wiki/SipHash).
* Review implementation in https://github.com/ClickHouse/ClickHouse/blob/c3a9cbe01c1c07d21cd1f89b31291a194fc21b25/src/Functions/FunctionsHashing.h#L204

**Expected behavior**
Add two new functions which allow for user-defined keys/seeds:
* `sipHash64Keyed(key,par1,...)` 
* `sipHash128Keyed(key,par1,...)`

### PR DESCRIPTION
**Pull request by aiven-sal**
Closes https://github.com/ClickHouse/ClickHouse/issues/22520

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add two new functions which allow for user-defined keys/seeds with SipHash{64,128}

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

### CODE REVIEW

File: docs/en/sql-reference/functions/hash-functions.md [hunk 1]
```diff
@@ -84,7 +84,39 @@ SELECT sipHash64(array('e','x','a'), 'mple', 10, toDateTime('2019-06-15 23:00:00
 
 ```response
 ┌──────────────SipHash─┬─type───┐
-│ 13726873534472839665 │ UInt64 │
```
Comment (qoega) [inline docs/en/sql-reference/functions/hash-functions.md:87]: This one looks suspicious, or we just had incorrect example?

File: src/Functions/FunctionsHashing.h [hunk 2]
```diff
@@ -896,17 +992,19 @@ class FunctionIntHash : public TargetSpecific::Default::FunctionIntHash<Impl, Na
...
+                  "can't use IntHash does not support keys");
 public:
     static constexpr auto name = Impl::name;
 
 private:
     using ToType = typename Impl::ReturnType;
 
     template <typename FromType, bool first>
-    void executeIntType(const IColumn * column, typename ColumnVector<ToType>::Container & vec_to) const
+    void executeIntType(const IColumn * column, typename ColumnVector<ToType>::Container & vec_to, const ColumnWithTypeAndName & key) const
```
Comment (rschu1ze) [inline src/Functions/FunctionsHashing.h:1007]: Looks a bit weird and also is not consistent with the semantics in SQL to add another input argument after the output argument. Suggest order: 1. key 2. column 3. vec_to.

File: src/Functions/FunctionsHashing.h [hunk 3]
```diff
@@ -1240,18 +1338,40 @@ class FunctionAnyHash : public IFunction
 
         typename ColumnVector<ToType>::Container & vec_to = col_to->getData();
 
-        if (arguments.empty())
+        if (Keyed)
         {
-            /// Constant random number from /dev/urandom is used as a hash value of empty list of arguments.
-            vec_to.assign(rows, static_cast<ToType>(0xe28dbde7fe22e41c));
+            // First argument will be used as key and it's mandatory
```
Comment (rschu1ze) [inline src/Functions/FunctionsHashing.h:1343]: Quite a lot of copypasta, could you somehow unify l. 1341 and l. 1361?

File: docs/en/sql-reference/functions/hash-functions.md [hunk 4]
```diff
@@ -84,7 +84,39 @@ SELECT sipHash64(array('e','x','a'), 'mple', 10, toDateTime('2019-06-15 23:00:00
...
+
+**Syntax**
+
+```sql
+sipHash64Keyed((k0, k1), par1,...)
+```
+
+**Arguments**
+
+The function takes a variable number of input parameters. The first argument is a tuple of two UInt64 representing the key. The other arguments can be any of the [supported data types](/docs/en/sql-reference/data-types/index.md). For some data types calculated value of hash function may be the same for the same values even if types of arguments differ (integers of different size, named and unnamed `Tuple` with the same data, `Map` and the corresponding `Array(Tuple(key, value))` type with the same data).
```
[the remark this answers is not in the document]
Comment (rschu1ze) [reply docs/en/sql-reference/functions/hash-functions.md:103]: The "keyed" variants ideally behave as consistent with the other hash functions as possible. So even if throwing an exception looks more appealing, we should rather return a magic number.

File: docs/en/sql-reference/functions/hash-functions.md [hunk 5]
```diff
@@ -84,7 +84,39 @@ SELECT sipHash64(array('e','x','a'), 'mple', 10, toDateTime('2019-06-15 23:00:00
...
+**Returned value**
+
+A [UInt64](/docs/en/sql-reference/data-types/int-uint.md) data type hash value.
+
+**Example**
+
+Query:
+
+```sql
+SELECT sipHash64Keyed((toUInt64(0), toUInt64(0)), array('e','x','a'), 'mple', 10, toDateTime('2019-06-15 23:00:00')) AS SipHash, toTypeName(SipHash) AS type;
```
[the remark this answers is not in the document]
Comment (rschu1ze) [reply docs/en/sql-reference/functions/hash-functions.md:114]: Ah, I see. Up to you ... you can also include both examples in the docs and add a comment behind the 0-example a la `-- equivalent to sipHash64()`.

**Comment by rschu1ze** [comment]
@aiven-sal I took the liberty of making some edits in the documentation. Looking at the actual code now.

File: src/Functions/FunctionsHashing.h [hunk 6]
```diff
@@ -274,6 +274,54 @@ struct SipHash64Impl
     static constexpr bool use_int_hash_for_pods = false;
 };
 
+struct SipHash64KeyedImpl
+{
+    static constexpr auto name = "sipHash64Keyed";
+    using ReturnType = UInt64;
+    struct KeyType
```
Comment (rschu1ze) [inline src/Functions/FunctionsHashing.h:281]: Minor cosmetic comment: Name `*Type` (or `*T`) is usually for generic or templated types, in this case, just `Key` would do the job.

File: src/Functions/FunctionsHashing.h [hunk 7]
```diff
@@ -274,6 +274,54 @@ struct SipHash64Impl
...
+    static KeyType parseKey(const ColumnWithTypeAndName & key)
+    {
+        KeyType ret;
+
+        const auto * tuple = checkAndGetColumn<ColumnTuple>(key.column.get());
+        if (!tuple)
+            throw Exception(ErrorCodes::NOT_IMPLEMENTED, "key must be a tuple");
+
+        if (tuple->tupleSize() != 2)
+            throw Exception(ErrorCodes::NOT_IMPLEMENTED, "wrong tuple size: key must be a tuple of 2 UInt64");
```
Comment (rschu1ze) [inline src/Functions/FunctionsHashing.h:296]: Totally minor, but the check at l. 295 only checks that the input is a 2-valued tuple and not the actual data types (l. 296). Maybe rephrase: "key argument must be a tuple with two elements"

File: src/Functions/FunctionsHashing.h [hunk 8]
```diff
@@ -274,6 +274,54 @@ struct SipHash64Impl
...
+{
+    static constexpr auto name = "sipHash64Keyed";
+    using ReturnType = UInt64;
+    struct KeyType
+    {
+        UInt64 key0 = 0;
+        UInt64 key1 = 0;
+    };
+
+    static KeyType parseKey(const ColumnWithTypeAndName & key)
```
Comment (rschu1ze) [inline src/Functions/FunctionsHashing.h:287]: I am fine with a static function in SipHash64KeyedImpl. Since SipHash128 also calls it (l. 350), we could have placed the code also into a "detail"/"impl" namespace (*), e.g. 

```cpp
namespace impl
{

static KeyType parseSipHashKey(const ColumnWithTypeAndName & key)
{
    ...
}

}
```

EDIT: Same could be done for the `KeyType` since that is shared with sipHash128 (l. 348)

```cpp
namespace impl
{

struct SipHashKey
{
...
};

}
```

(*) Normally, anonymous namespaces are used for that but they are not recommended in headers.

File: src/Functions/FunctionsHashing.h [hunk 9]
```diff
@@ -1240,17 +1310,31 @@ class FunctionAnyHash : public IFunction
 
         typename ColumnVector<ToType>::Container & vec_to = col_to->getData();
 
-        if (arguments.empty())
+        /// If using a "keyed" algorithm, the first argument is the key and
+        /// the data starts from the second argument.
+        /// Otherwise there is no key and all arguments are interpreted as data.
+        constexpr size_t first_data_argument = Keyed;
+
+        if (arguments.size() <= first_data_argument)
```
Comment (rschu1ze) [inline src/Functions/FunctionsHashing.h:1318]: (It has been like that before but the comment in l. 1320 is confusing ... it should just say that we produce a fixed, random-looking magic value if the input is empty)

File: src/Functions/FunctionsHashing.h [hunk 10]
```diff
@@ -1240,17 +1310,31 @@ class FunctionAnyHash : public IFunction
...
         {
             /// Constant random number from /dev/urandom is used as a hash value of empty list of arguments.
             vec_to.assign(rows, static_cast<ToType>(0xe28dbde7fe22e41c));
         }
 
-        /// The function supports arbitrary number of arguments of arbitrary types.
+        KeyType key{};
+        if constexpr (Keyed)
```
Comment (rschu1ze) [inline src/Functions/FunctionsHashing.h:1325]: In general, feel free to omit parentheses for single-line if (constexpr) statements, e.g.
```cpp
if constexpr (Keyed)
    if (!arguments.empty())
        key = Impl::parseKey(arguments[0]);
```

File: src/Functions/FunctionsHashing.h [hunk 11]
```diff
@@ -1240,17 +1310,31 @@ class FunctionAnyHash : public IFunction
...
+        {
+            if (!arguments.empty())
+                key = Impl::parseKey(arguments[0]);
+        }
 
+        /// The function supports arbitrary number of arguments of arbitrary types.
         bool is_first_argument = true;
-        for (const auto & col : arguments)
-            executeForArgument(col.type.get(), col.column.get(), vec_to, is_first_argument);
+        for (std::size_t i = first_data_argument; i < arguments.size(); ++i)
```
Comment (rschu1ze) [inline src/Functions/FunctionsHashing.h:1333]: `std::size_t` --> `size_t`

**Review by rschu1ze — APPROVED** [review]
Thanks for incorporating the feedback. I just have some cosmetic/minor comments. Approval is granted, if the tests look good I'll merge.

File: src/Functions/FunctionsHashing.h (hunk 7 above)
Comment (rschu1ze) [reply src/Functions/FunctionsHashing.h:296]: You can do it how you like, any alternative is fine with me.

**Review by rschu1ze — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
@aiven-sal, if no one explains what is happening here: https://github.com/ClickHouse/ClickHouse/pull/50245#issuecomment-1566360561, we will have to revert this pull request.

**Comment by alexey-milovidov** [comment]
We have to revert it.
